ppb_vector/vector2.py

```diff
@@ -1,97 +1,174 @@
 import typing
 import collections
+import functools
 from math import acos, atan2, cos, degrees, hypot, isclose, radians, sin
 from numbers import Real
 from collections.abc import Sequence
 
 __all__ = 'Vector2',
 
 
+# Vector or subclass
+VectorOrSub = typing.TypeVar('VectorOrSub', bound='Vector2')
+
+Realish = typing.Union[Real, float, int]
+
+# Anything convertable to a Vector, including lists, tuples, and dicts
 VectorLike = typing.Union[
-    'Vector2',
-    typing.List[Real],  # TODO: Length 2
-    typing.Tuple[Real, Real],
-    typing.Dict[str, Real],  # TODO: Length 2, keys 'x', 'y'
+    'Vector2',  # Or subclasses, unconnected to the VectorOrSub typevar above
+    typing.List[typing.SupportsFloat],  # TODO: Length 2
+    typing.Tuple[typing.SupportsFloat, typing.SupportsFloat],
+    typing.Dict[str, typing.SupportsFloat],  # TODO: Length 2, keys 'x', 'y'
 ]
 
 
 def is_vector_like(value: typing.Any) -> bool:
     return isinstance(value, (Vector2, list, tuple, dict))
 
 
-_fakevector = collections.namedtuple('_fakevector', ['x', 'y'])
+@functools.lru_cache()
+def _find_lowest_type(left: typing.Type, right: typing.Type) -> typing.Type:
+    """
+    Guess which is the more specific type.
+    """
+    # Basically, see what classes are unique in each type's MRO and return who
+    # has the most.
+    lmro = set(left.__mro__)
+    rmro = set(right.__mro__)
+    lspecial = lmro - rmro
+    rspecial = rmro - lmro
+    if len(lmro) > len(rmro):
+        return left
+    elif len(rmro) > len(lmro):
+        return right
+    else:
+        # They're equal, just arbitrarily pick one
+        return left
+
 
-def _mkvector(value, *, castto=_fakevector):
-    if isinstance(value, Vector2):
-        return value
-    # FIXME: Allow all types of sequences
-    elif isinstance(value, (list, tuple)) and len(value) == 2:
-        return castto(value[0], value[1])
-    # FIXME: Allow all types of mappings
-    elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
-        return castto(value['x'], value['y'])
+def _find_lowest_vector(left: typing.Type, right: typing.Type) -> typing.Type:
+    if left is right:
+        return left
+    elif not issubclass(left, Vector2):
+        return right
+    elif not issubclass(right, Vector2):
+        return left
     else:
-        raise ValueError(f"Cannot use {value} as a vector-like")
+        return _find_lowest_type(left, right)
 
 
-class Vector2(Sequence):
+class Vector2:
+    x: float
+    y: float
+    _length: float
 
-    def __init__(self, x: Real, y: Real):
-        self.x = x
-        self.y = y
-        self.length = hypot(x, y)
+    def __init__(self, x: typing.SupportsFloat, y: typing.SupportsFloat):
+        try:
+            self.x = x.__float__()
```

Comment:
Why is this calling the dunder, rather than using `float()` ?

Reply:
Disallowing strings.

Reply:
OK, sounds good <3